### PR TITLE
[7/7] Multimodal datasets (The Cauldron, LLaVA-Instruct-150K)

### PR DESCRIPTION
**Pull request by RdoubleA**
## Changelog
- Add two example multimodal dataset builders, each with their own dataset transform: [The Cauldron](https://huggingface.co/datasets/HuggingFaceM4/the_cauldron) and [LLaVA-Instruct-150K](https://huggingface.co/datasets/liuhaotian/LLaVA-Instruct-150K). Both require slightly different preprocessing and serve as good examples
- Add a utility to quickly map raw text with image tags into what's expected for Message content field
- Upgrade `ShareGPTToMessages` with support for an image column

## Test plan
- Unit test for The Cauldron, LLaVA Instruct datasets
- Unit test for split_text_by_image_tag

### CODE REVIEW

File: torchtune/data/_utils.py
Comment (ebsmothers) [inline torchtune/data/_utils.py:38]: Can you be a bit more explicit about the format? E.g. at least "image tags will be represented as a dictionary {"type": "image"}"

File: torchtune/data/_utils.py
Comment (ebsmothers) [inline torchtune/data/_utils.py:58]: So this will not support multiple consecutive image tags, right? If so that should be explicitly stated somewhere

File: torchtune/data/_messages.py
Comment (ebsmothers) [inline torchtune/data/_messages.py:407]: Maybe controversial but why do we need to put all this in the existing `ShareGPTToMessages` class? Then as a text-only user I have extra stuff to sort through, plus when we expose at the dataset level it feels kinda overloaded. Idk maybe there are obvious pros to doing it this way that I'm missing here (apart from the obvious one of less code duplication, which tbh I don't care too much about in this case)

File: torchtune/datasets/multimodal/_llava_instruct.py
Comment (ebsmothers) [inline torchtune/datasets/multimodal/_llava_instruct.py:51]: This is very general and probably the most important part of setting this up correctly. I would try to give an explicit example here or something

File: torchtune/datasets/multimodal/_llava_instruct.py
Comment (ebsmothers) [inline torchtune/datasets/multimodal/_llava_instruct.py:66]: When would one of these not contain images though?

File: torchtune/datasets/multimodal/_the_cauldron.py
Comment (ebsmothers) [inline torchtune/datasets/multimodal/_the_cauldron.py:124]: I think I missed the original PR here, but this naming feels a bit unclear to me (though I guess the purpose of "new" is to emphasize that we're overriding any other system prompts). Also again would give an example of this

File: torchtune/datasets/multimodal/_the_cauldron.py
Comment (ebsmothers) [inline torchtune/datasets/multimodal/_the_cauldron.py:149]: Should be passing `new_system_prompt`? Also nit but would call it `to_messages` or something, since `model_transform` also involves `Message` but as an input

File: torchtune/datasets/multimodal/_the_cauldron.py
Comment (ebsmothers) [inline torchtune/datasets/multimodal/_the_cauldron.py:84]: Based on this I take it Cauldron never has its own system prompt?

File: torchtune/datasets/multimodal/_the_cauldron.py
Comment (ebsmothers) [inline torchtune/datasets/multimodal/_the_cauldron.py:22]: Give examples please

File: torchtune/datasets/multimodal/_the_cauldron.py
Comment (ebsmothers) [inline torchtune/datasets/multimodal/_the_cauldron.py:49]: nit: this wording is unclear

File: torchtune/datasets/multimodal/_the_cauldron.py
Comment (ebsmothers) [inline torchtune/datasets/multimodal/_the_cauldron.py:15]: I do worry a bit that we're overfitting to these two datasets in our choice of naming. Like really iiuc llava_instruct means you have a local directory of images and the sample points to filenames and the cauldron means you have actual PILs in each sample. That seems more general than what we're indicating

File: torchtune/datasets/multimodal/_the_cauldron.py
Comment (ebsmothers) [inline torchtune/datasets/multimodal/_the_cauldron.py:90]: Maybe I'm being dense but looking at this I have no idea how I should define the model_transform to construct one of these

**Review by ebsmothers — COMMENTED** [review]
Need to fix CI

File: torchtune/data/_messages.py
Comment (RdoubleA) [reply torchtune/data/_messages.py:407]: It seemed general enough to include in a general transform, but yeah I can see this being overloaded. Some options are: 1) make a separate ShareGPT transform with images, 2) keep the llava instruct transform its own class until we find more examples that fit this format (don't generalize until we need to ;) )

File: torchtune/data/_utils.py
Comment (RdoubleA) [reply torchtune/data/_utils.py:58]: hmmm.... it should. let me revisit

File: torchtune/datasets/multimodal/_llava_instruct.py
Comment (RdoubleA) [reply torchtune/datasets/multimodal/_llava_instruct.py:51]: once the flamingo transform lands, it would be easy to point to this

File: torchtune/datasets/multimodal/_llava_instruct.py
Comment (RdoubleA) [reply torchtune/datasets/multimodal/_llava_instruct.py:66]: mmm... I guess you're right, I suppose it doesn't make sense to make this configurable

File: torchtune/datasets/multimodal/_the_cauldron.py
Comment (RdoubleA) [reply torchtune/datasets/multimodal/_the_cauldron.py:15]: well, we only have one dataset example that fits this mold. so does it make sense to make it general at this point?

File: torchtune/datasets/multimodal/_the_cauldron.py
Comment (RdoubleA) [reply torchtune/datasets/multimodal/_the_cauldron.py:84]: the cauldron doesn't by default but we'd want it to be configurable if the user wants to add it

File: torchtune/datasets/multimodal/_the_cauldron.py
Comment (RdoubleA) [reply torchtune/datasets/multimodal/_the_cauldron.py:124]: Yep that's exactly it

File: torchtune/datasets/multimodal/_llava_instruct.py
Comment (ebsmothers) [reply torchtune/datasets/multimodal/_llava_instruct.py:51]: Can we add a TODO or a follow-up then? I wanna make sure it's very clear how to use these builders, rn I feel like it's not

File: torchtune/data/_messages.py
Comment (ebsmothers) [reply torchtune/data/_messages.py:407]: OK yeah let's do (2). Plus we already have a Cauldron transform so I feel like it's most consistent to do it this way anyways

File: torchtune/datasets/multimodal/_the_cauldron.py
Comment (ebsmothers) [reply torchtune/datasets/multimodal/_the_cauldron.py:15]: Yeah I guess this is kinda inconsistent with my other comment anyways. Let's go with what you proposed there, keep the naming as is and have a separate transform for each of the cauldron and llava-instruct datasets

File: torchtune/datasets/multimodal/_llava_instruct.py
Comment (ebsmothers) [inline torchtune/datasets/multimodal/_llava_instruct.py:189]: This part is a bit confusing to me. So model transform has a max_seq_len, but that's really just associated with the tokenizer, right? So do we pass that up from the tokenizer to the model transform (since it's composed of both tokenizer and image transform)?

File: torchtune/datasets/multimodal/_the_cauldron.py
Comment (ebsmothers) [inline torchtune/datasets/multimodal/_the_cauldron.py:142]: Sorry to harp on this, but can we have a more explicit example until the Flamingo transform is available? Even if it's just a mock tokenizer and image transform with clearly-defined API contracts in a code snippet, I think that will help a lot

File: torchtune/data/_messages.py
Comment (ebsmothers) [inline torchtune/data/_messages.py:468]: If we're moving this here, should also delete from data/_utils.py?

Comment (ebsmothers) [reply torchtune/data/_messages.py:468]: Also for multimodal, do we need to do any extra validation? E.g. number of images == number of messages with type == 'image'? (Not saying we should do it here btw)

**Review by ebsmothers — APPROVED** [review]
A few more comments but no major concerns from me. Stamping to unblock

File: tests/torchtune/datasets/multimodal/test_llava_instruct_dataset.py
Comment (RdoubleA) [reply tests/torchtune/datasets/multimodal/test_llava_instruct_dataset.py:74]: image token, eos id

File: torchtune/datasets/multimodal/_llava_instruct.py
Comment (RdoubleA) [reply torchtune/datasets/multimodal/_llava_instruct.py:118]: oop

File: torchtune/data/_messages.py
Comment (RdoubleA) [reply torchtune/data/_messages.py:468]: good point... I am not sure yet where this should happen. maybe note this as a followup once the whole end-to-end is established

File: torchtune/datasets/multimodal/_llava_instruct.py
Comment (RdoubleA) [reply torchtune/datasets/multimodal/_llava_instruct.py:189]: excellent question, @pbontrager and I were considering enforcing some contract at the model transform level, but maybe for now we just pass that up from the tokenizer

**Pull request merged**
